dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.common.dimension;
+
+import static java.util.Arrays.stream;
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.StringUtils.split;
+import static org.hisp.dhis.analytics.common.dimension.DimensionIdentifier.ElementWithOffset.emptyElementWithOffset;
+
+import java.util.List;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor( access = AccessLevel.PRIVATE )
+public class DimensionIdentifierConverterSupport
+{
+
+    private static final Character DIMENSION_SEPARATOR = '.';
+
+    public static DimensionIdentifier<StringUid, StringUid, StringUid> fromFullDimensionId( String fullDimensionId )
+    {
+        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets = parseFullDimensionId( fullDimensionId );
+
```

Comment:
It is ok, but what about try this:
```
public static DimensionIdentifier<StringUid,` StringUid, StringUid> fromFullDimensionId( String fullDimensionId )
    {
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets = parseFullDimensionId( fullDimensionId );

        return DimensionIdentifier.of( getProgram(uidWithOffsets), getStage(uidWithOffsets),
                getDimension(uidWithOffsets) );
    }

```
parseFullDimensionId method can throw 
        throw new IllegalArgumentException( "Malformed parameter: " + fullDimensionId );

Reply:
An alternative would look like this:

```java
    public static DimensionIdentifier<StringUid, StringUid, StringUid> fromFullDimensionId( String fullDimensionId )
    {
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets = parseFullDimensionId( fullDimensionId );

        return DimensionIdentifier.of(
            getProgram( uidWithOffsets ),
            getProgramStage( uidWithOffsets ),
            getDimension( uidWithOffsets ) );
    }

    private static DimensionIdentifier.ElementWithOffset<StringUid> getProgram(
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets )
    {
        if ( uidWithOffsets.size() == 2 || uidWithOffsets.size() == 3 )
        {
            return uidWithOffsets.get( 0 );
        }
        else if ( uidWithOffsets.size() == 1 )
        {
            return emptyElementWithOffset();
        }
        throw new IllegalArgumentException( "Invalid dimension identifier: " + uidWithOffsets );
    }

    private static DimensionIdentifier.ElementWithOffset<StringUid> getProgramStage(
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets )
    {
        if ( uidWithOffsets.size() == 3 )
        {
            return uidWithOffsets.get( 1 );
        }
        else if ( uidWithOffsets.size() == 2 || uidWithOffsets.size() == 1 )
        {
            return emptyElementWithOffset();
        }
        else
        {
            throw new IllegalArgumentException( "Invalid dimension identifier: " + uidWithOffsets );
        }
    }

    private static StringUid getDimension( List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets )
    {
        DimensionIdentifier.ElementWithOffset<StringUid> dimension;
        if ( uidWithOffsets.size() == 3 )
        {
            dimension = uidWithOffsets.get( 2 );
        }
        else if ( uidWithOffsets.size() == 2 )
        {
            dimension = uidWithOffsets.get( 1 );
        }
        else if ( uidWithOffsets.size() == 1 )
        {
            dimension = uidWithOffsets.get( 0 );
        }
        else
        {
            throw new IllegalArgumentException( "Invalid dimension identifier: " + uidWithOffsets );
        }
        assertDimensionIdHasNoOffset( dimension );
        return dimension.getElement();
    }
```